src/utility/copy.cpp

```diff
@@ -59,12 +70,25 @@ CreateRelationCopyParseState(ParseState *pstate, const CopyStmt *stmt, List **va
 		    lappend(*vars, makeVar(1, cur, tuple_desc->attrs[cur - 1].atttypid, tuple_desc->attrs[cur - 1].atttypmod,
 		                           tuple_desc->attrs[cur - 1].attcollation, 0));
 	}
+#else
+	foreach(cur, attnums)
+	{
+		int	attno = lfirst_int(cur) - FirstLowInvalidHeapAttributeNumber;
```

Comment:
Does this work? That should allow us to remove the `ListCell` declaration at the top of the function. If not we should update these macros to work for PG15.

```suggestion
	foreach_int(cur, attnums)
	{
		int	attno = cur - FirstLowInvalidHeapAttributeNumber;
```

Reply:
This part refers to part of the source code of PG15
![image](https://github.com/user-attachments/assets/a16c4c65-ecb4-4cdd-84b3-948e0901ed62)

Reply:
Hmm, I didn't realize that we copied a part of this code initially. I might have kept the `foreach` macro then when I originally replaced it. But let's stay consistent here for now and use the new `foreach_int` macro in both places.

Reply:
I have just read the relevant code and understood your meaning. Try to use the relevant definition of "pg_list.hpp" in the project, and I will modify it. 
```c
#define foreach_ptr(type, var, lst) foreach_internal(type, *, var, lst, lfirst)
#define foreach_int(var, lst)	foreach_internal(int, , var, lst, lfirst_int)
#define foreach_oid(var, lst)	foreach_internal(Oid, , var, lst, lfirst_oid)
#define foreach_xid(var, lst)	foreach_internal(TransactionId, , var, lst, lfirst_xid)

/*
 * The internal implementation of the above macros.  Do not use directly.
 *
 * This macro actually generates two loops in order to declare two variables of
 * different types.  The outer loop only iterates once, so we expect optimizing
 * compilers will unroll it, thereby optimizing it away.
 */
#define foreach_internal(type, pointer, var, lst, func) \
	for (type pointer var = 0, pointer var##__outerloop = (type pointer) 1; \
		 var##__outerloop; \
		 var##__outerloop = 0) \
		for (ForEachState var##__state = {(lst), 0}; \
			 (var##__state.l != NIL && \
			  var##__state.i < var##__state.l->length && \
			 (var = func(&var##__state.l->elements[var##__state.i]), true)); \
			 var##__state.i++)
```

Reply:
Sorry, using part of PG code has become a habit, I did not understand your meaning quickly